lib/handler/cache-handler.js

```diff
@@ -0,0 +1,368 @@
+'use strict'
+
+const util = require('../core/util')
+const DecoratorHandler = require('../handler/decorator-handler')
+const { parseCacheControlHeader, parseVaryHeader, UNSAFE_METHODS } = require('../util/cache')
+
+/**
+ * Writes a response to a CacheStore and then passes it on to the next handler
+ */
+class CacheHandler extends DecoratorHandler {
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.CacheOptions} | null
+   */
+  #opts = null
+  /**
+   * @type {import('../../types/dispatcher.d.ts').default.RequestOptions | null}
+   */
+  #req = null
+  /**
+   * @type {import('../../types/dispatcher.d.ts').default.DispatchHandlers | null}
+   */
+  #handler = null
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.CacheStoreWriteable | undefined}
+   */
+  #writeStream
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions} opts
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   */
+  constructor (opts, req, handler) {
+    super(handler)
+
+    this.#opts = opts
+    this.#req = req
+    this.#handler = handler
+  }
+
+  /**
+   * @see {DispatchHandlers.onHeaders}
+   *
+   * @param {number} statusCode
+   * @param {Buffer[]} rawHeaders
+   * @param {() => void} resume
+   * @param {string} statusMessage
+   * @param {Record<string, string | string[]> | undefined} headers
+   * @returns {boolean}
+   */
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = util.parseHeaders(rawHeaders)
+  ) {
+    if (
+      this.#req.method in UNSAFE_METHODS &&
+      statusCode >= 200 &&
+      statusCode <= 399
+    ) {
+      // https://www.rfc-editor.org/rfc/rfc9111.html#name-invalidating-stored-respons
+      const result = this.#opts.store.deleteByOrigin(this.#req.origin)
+      if (
+        result &&
+        typeof result.catch === 'function' &&
+        typeof this.#handler.onError === 'function'
+      ) {
+        result.catch(err => {
+          this.#handler.onError(err)
+        })
+      }
+
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlHeader = headers['cache-control']
+    const contentLengthHeader = headers['content-length']
+
+    if (!cacheControlHeader || !contentLengthHeader || this.#opts.store.isFull) {
+      // Don't have the headers we need, can't cache
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
```

Comment:
```suggestion
      return this.#handler.onHeaders(
        statusCode,
        rawHeaders,
        resume,
        statusMessage,
        headers
      )
```

This same part is replicated multiple times in this function. I think there is room for refactoring/simplification?